File: 프로그래머스/LV3/정수 삼각형.py
def solution(triangle):
    dp = [[0] * (i + 1) for i in range(len(triangle))]
    dp[0][0] = triangle[0][0]

    for i in range(len(triangle) - 1):
        for j in range(len(triangle[i])):
            dp[i + 1][j + 1] = dp[i][j] + triangle[i + 1][j + 1]
            if dp[i + 1][j] == 0:
                dp[i + 1][j] = dp[i][j] + triangle[i + 1][j]
            else:
                dp[i + 1][j] = max(dp[i + 1][j], dp[i][j] + triangle[i + 1][j])

    return max(dp[len(dp) - 1])

File: 프로그래머스/javascript/LV3/정수삼각형.js
function solution(triangle) {
  const dp = Array.from({ length: triangle.length }, (_, i) =>
    Array(i + 1).fill(0),
  );
  dp[0][0] = triangle[0][0];

  for (let i = 1; i < triangle.length; i++) {
    for (let j = 0; j < triangle[i].length; j++) {
      if (j === 0) {
        dp[i][j] = dp[i - 1][0] + triangle[i][j];
      } else if (j === triangle[i].length - 1) {
        dp[i][j] = dp[i - 1][j - 1] + triangle[i][j];
      } else {
        dp[i][j] = Math.max(
          dp[i - 1][j - 1] + triangle[i][j],
          dp[i - 1][j] + triangle[i][j],
        );
      }
    }
  }
  return Math.max(...dp[dp.length - 1]);
}
